### ISSUE TITLE
Remove python from the coreclr build

### ISSUE THREAD

**Issue by jashook** [dotnet/runtime#38574]
Currently this is only used in a few places. We can easily translate the scripts which use it to either `powershell` or `msbuild` to drop the build dependency.

**Comment by hoyosjs** [comment]
Given that the generation also needs to work in the native build, msbuild and powershell might not be the best candidates. How often do the generated sources change? (or how often do the inputs change)

@sywhang @josalem

**Comment by wfurt** [comment]
This may increase difficulty of porting to new OSes and architectures where one could build natives parts without .NET presence and build managed assemblies on supported platforms. 
@am11 @janvorli

**Comment by am11** [comment]
Agreed. Considering python scripts which are responsible to run coreclr tests; from portability viewpoint:

* powershell is not a good candidate because in addition to the set of dependencies required to build dotnet/runtime, it depends on OMI and other native components which are not available on all platforms.
* msbuild is not a good candidate because it depends on parts of class libraries, which might not be implemented on target system at the time when developer needs to run tests. (especially System.Diagnostics.Process and System.IO.File could be the tough ones depending on the platform..)

Note that today we need to even patch those python scripts to skip calling SDK (which is used for determining whether or not to rebuild the test project); in order to run the tests from pre- cross-compiled packages and assume there is no SDK or auxiliary dotnet library present. I think this is something which can be properly implemented in runtest.py and friends to not look for SDK at all. If we instead switch to  MSBuild for running tests, then it will simply become impossible to run coreclr tests on a new platform, before porting all libraries that SDK and MSBuild depend on.

**Comment by janvorli** [comment]
I also don't like moving to powershell or msbuild because of the reason that @am11 mentioned.

**Comment by juliusfriedman** [comment]
Just use the system shell, what complex works needs to be done? 

stop parsing YML and just do things like normal people...

**Comment by danmosemsft** [comment]
@juliusfriedman  unlike shell scripts, Python works on all platforms.

Please be careful with your tone.

**Comment by juliusfriedman** [comment]
> @juliusfriedman unlike shell scripts, Python works on all platforms.
> 
> Please be careful with your tone.

Look at the other comments in this thread, compared to mine what did I say that was so bad?

Please, Show me 1 script I can't rewrite using pure cmd or bash and I will `watch` my tone, rather observe the extent of my written words such they hopefully not be interpreted with tone as you have.

I bet its the YML related ones and I bet I could make something to split them up easily enough... [here](https://stackoverflow.com/questions/2524928/dos-batch-iterate-through-a-delimited-string)... I think we want new lines up to the point we reach the token which was at line 1 or the last parsed line for the YML 📦

**Comment by ViktorHofer** [comment]
Posting @jkotas' comment about the python dependency from another thread: https://github.com/dotnet/runtime/pull/39052#issuecomment-656709021

> For bootstrapping build, we can have a file with empty event firing stubs checked in (or we can be a very trivial script written to generate this). And once we are fully bootstrapped, the full eventing code generator can be written in C#.

> The regular coreclr test execution should not need python. It is using xunit as the driver.
We can have the python script to run tests around for bootstrapping, but we do not need to be forcing the python dependency on everybody for that.

**Comment by huoyaoyuan** [comment]
I'm interested to look into this.

For the eventing case I have some questions. Python is used for generating cpp stubs to fire events. Will native build run again with managed SDK bootstrapped? Is there any way to detect it?

**Comment by am11** [comment]
@huoyaoyuan it's discussed in this thread and see the reaction of the owners: https://github.com/dotnet/runtime/issues/57784#issuecomment-946116352. I have a few better ideas about it since then incorporating their concerns, one of them is looking promising (uses cmake without any external tool).

**Comment by am11** [comment]
@huoyaoyuan it's discussed in another thread and see the reaction of the owners: https://github.com/dotnet/runtime/issues/57784#issuecomment-946116352. I have a few better ideas about it since then incorporating their concerns, one of them is looking promising (uses cmake without any external tool).

**Comment by huoyaoyuan** [comment]
> input: we define a static header with our events and groups and keep it in sync with `src/coreclr/vm/ClrEtwAll.man` (manually at first, as it's only a few times a year we add/modify events; if we have more time, we can use XmlPeek msbuild task to update the header files when msbuild is present .. because cmake doesn't support xml parsing)

Yeah, this is exactly what I'm considering. The xml parsed result for `ClrEtwAll.man` can be checked in with macro-friendly format, and cpp sources can use macros. The file could be re-generated when msbuild is available, or dry-run in CI to avoid forgetting to check-in.

The referenced thread looks focusing on another topic about LTTNG versioning. I think the generation mechanism change can be orthogonal to that.

**Comment by am11** [comment]
It is about generating it from python, that's where I did my experiment with XLTS transformation etc. I have been waiting for a few things in cmake which will make this translation very smooth. Lets not rush here..

**Comment by am11** [comment]
It is about generating it from python, that's where I did my experiment with XSLT transformation etc. I have been waiting for a few things in cmake which will make this translation very smooth. Lets not rush here..

**Comment by huoyaoyuan** [comment]
> I have been waiting for a few things in cmake which will make this translation very smooth.

Any more context here? I can't find anything related to xml parsing in cmake.

Indeed there are two strategies: parse the xml in native tool available during bootstrap (probably cmake), or check-in the parsed result by other tool available. I can't see the former in foreseeable future. Even if it's available, we still have to wait Linux distros to adopt it.

**Comment by am11** [comment]
> or check-in the parsed result by other tool available

I had its PoC working like four years ago with XSLT used in msbuild task, but then read the area owner want it to run every time including when msbuild is not available, so I moved on to do something else.

Right away, there is no single component which can cover it.

There is a [feature request](https://gitlab.kitware.com/cmake/cmake/-/issues/26097), which is only a few weeks old.

What we have at our disposal today is `string(JSON)` support in the minimum version of cmake required by dotnet/runtime across the board. If you don't want to wait for `string(XML)`, a future-proof implementation could be to create `eng/transform-clretw.proj` that does the xml->json STJ thing withing a code fragment in that proj file (without necessarily having .cs/.csproj or other ceremony), and run it to convert ClrEtwAll.man to ClrEtwAll.man.json. That will only run when msbuild is available (which is most of the times) and get skipped when directly invoking internal scripts like src/clr/build-runtime etc. We have several examples of "dummy" moving parts in that build mode like cdac and dummy event provider etc. anyway, so whatever state of ClrEtwAll.man.json would be there by the time cmake is invoked from gen-buildsys script, it will just use it to write the headers with macros.

**Comment by am11** [comment]
> the minimum version of cmake required by dotnet/runtime across the board

I thought https://github.com/dotnet/runtime/pull/103088 was merged. So read it "after #103088 is merged" 😅

**Comment by huoyaoyuan** [comment]
> I had its PoC working like four years ago with XSLT used in msbuild task, but then read the area owner want it to run every time including when msbuild is not available, so I moved on to do something else.

I see the rationale on this. But I really want to make python unavailable in my environments. It's not a default of Windows either.

> There is a [feature request](https://gitlab.kitware.com/cmake/cmake/-/issues/26097), which is only a few weeks old.

Well it's still far from being realistic.

Anyway, the approach on this should be up to decision for area owners. I've got the context around this. Thanks.
